pkgs/build-support/fetchgitlab/default.nix

```diff
@@ -1,32 +1,62 @@
-{ fetchgit, fetchzip, lib }:
+{ lib, fetchgit, fetchzip }:
 
 # gitlab example
 { owner, repo, rev, protocol ? "https", domain ? "gitlab.com", name ? "source", group ? null
-, fetchSubmodules ? false, leaveDotGit ? false, deepClone ? false
+, fetchSubmodules ? false, leaveDotGit ? false
+, deepClone ? false, forceFetchGit ? false
+, sparseCheckout ? []
+, private ? false, varPrefix ? null
 , ... # For hash agility
 } @ args:
 
 let
   slug = lib.concatStringsSep "/" ((lib.optional (group != null) group) ++ [ owner repo ]);
   escapedSlug = lib.replaceStrings [ "." "/" ] [ "%2E" "%2F" ] slug;
   escapedRev = lib.replaceStrings [ "+" "%" "/" ] [ "%2B" "%25" "%2F" ] rev;
-  passthruAttrs = removeAttrs args [ "protocol" "domain" "owner" "group" "repo" "rev" "fetchSubmodules" "leaveDotGit" "deepClone" ];
+  passthruAttrs = removeAttrs args [ "protocol" "domain" "owner" "group" "repo" "rev" "fetchSubmodules" "forceFetchGit" "private" "varPrefix" "leaveDotGit" "deepClone" ];
 
-  useFetchGit = deepClone || fetchSubmodules || leaveDotGit;
+  varBase = "NIX${lib.optionalString (varPrefix != null) "_${varPrefix}"}_GITLAB_PRIVATE_";
+  useFetchGit = fetchSubmodules || leaveDotGit || deepClone || forceFetchGit || !(sparseCheckout == [] || sparseCheckout == "");
   fetcher = if useFetchGit then fetchgit else fetchzip;
 
+  privateAttrs = lib.optionalAttrs private (
+    lib.throwIfNot (protocol == "https")
+    "private token login is only supported for https" {
+      netrcPhase = ''
+        if [ -z "''$${varBase}USERNAME" -o -z "''$${varBase}PASSWORD" ]; then
+          echo "Error: Private fetchFromGitLab requires the nix building process (nix-daemon in multi user mode) to have the ${varBase}USERNAME and ${varBase}PASSWORD env vars set." >&2
+          exit 1
+        fi
+      '' + (if useFetchGit then ''
+        cat > netrc <<EOF
+        machine ${domain}
+                login ''$${varBase}USERNAME
+                password ''$${varBase}PASSWORD
+        EOF
+      '' else ''
+        # needed because fetchurl always sets --netrc-file if a netrcPhase is present
+        touch netrc
+
+        cat > private-token <<EOF
+        PRIVATE-TOKEN: ''$${varBase}PASSWORD
+        EOF
+        curlOpts="$curlOpts --header @./private-token"
```

Comment:
Can we not express this in a netrc file?

Reply:
GitLab can use HTTP Basic Authentication and a `.netrc` file *only* when used via Git (e.g. `forceFetchGit = true`).

https://docs.gitlab.com/ee/user/project/settings/project_access_tokens.html#project-access-tokens

> Use a project access token to authenticate:
>
>  *  With the [GitLab API](https://docs.gitlab.com/ee/api/index.html#personalprojectgroup-access-tokens).
>  *  With Git, when using HTTP Basic Authentication, use:
>     *   Any non-blank value as a username.
>     *   The project access token as the password.
>
> Project access tokens are similar to [group access tokens](https://docs.gitlab.com/ee/user/group/settings/group_access_tokens.html) and [personal access tokens](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html).

It needs a custom `PRIVATE-TOKEN` header for access via the GitLab API (`fetchzip`/`fetchurl`):
https://docs.gitlab.com/ee/api/#personalprojectgroup-access-tokens

I added a short clarification to the code and to the commit message.